framework/src/utils/SlepcSupport.C

```diff
@@ -0,0 +1,257 @@
+
+    // We are solving a Non-Hermitian eigenvalue problem by default
+    MooseEnum eigen_problem_type("HEP NHEP GHEP GNHEP GHIEP");
+    params.addParam<MooseEnum>("eigen_problem_type", eigen_problem_type,
+                               "HEP: Hermitian "
+                               "NHEP: Non-Hermitian "
+                               "GHEP: Gerneralized Hermitian "
+                               "GNHEP: Generalized Non-Hermitian "
+                               "GHIEP: Generalized indefinite Hermitian ");
+
+
+    // Which eigenvalues are interest
+    MooseEnum which_eigen_pairs("LM SM LR SR LI SI TM TR TI ALL");
+    params.addParam<MooseEnum>("which_eigen_pairs", which_eigen_pairs,
+                               "LM: largest magnitude "
+                               "SM: smallest magnitude "
+                               "LR: largest real "
+                               "SR: smallest real "
+                               "LI: largest imaginary "
+                               "SI: smallest imaginary "
+                               "TM: target magnitude "
+                               "TR: target real "
+                               "TI: target imaginary "
+                               "ALL: all eigenvalues ");
+   return params;
+  }
+
+  void
+  storeSlepcEigenProblemOptions(EigenProblem & eigen_problem, const InputParameters & params)
+  {
+    // eigen problem type
+    if (params.isParamValid("eigen_problem_type"))
+    {
+      // Extract the eigen problem type
+      const std::string & eigen_problem_type = params.get<MooseEnum>("eigen_problem_type");
+      eigen_problem.solverParams()._eigen_problem_type = Moose::stringToEnum<Moose::EigenProblemType>(eigen_problem_type);
+    }
+    // which eigen pairs
+    if (params.isParamValid("which_eigen_pairs"))
+    {
+      // which eigen pairs
+      const std::string & which_eigen_pairs = params.get<MooseEnum>("which_eigen_pairs");
+      eigen_problem.solverParams()._which_eigen_pairs = Moose::stringToEnum<Moose::WhichEigenPairs>(which_eigen_pairs);
+    }
+  }
+
+  void
+  storeSlepcOptions(FEProblemBase & fe_problem, const InputParameters & params)
+  {
+    // if it is not an eigenvalue  problem, do nothing and return.
+    if (!(dynamic_cast<EigenProblem *>(&fe_problem)))
+      return;
+
+    // can not use solve_type for eigenvalue problems
+    if (params.isParamValid("solve_type"))
+    {
+      mooseError("Can not use \"solve_type\" for eigenvalue problems, please use \"eigen_solve_type\" instead \n");
+    }
+    // eigen solve type
+    if (params.isParamValid("eigen_solve_type"))
+    {
+      // Extract the eigen solve type
+      const std::string & eigen_solve_type = params.get<MooseEnum>("eigen_solve_type");
+      fe_problem.solverParams()._eigen_solve_type = Moose::stringToEnum<Moose::EigenSolveType>(eigen_solve_type);
+    }
+  }
+
+  void
+  setEigenProblemOptions(SolverParams & solver_params)
+  {
+    // set eigen problem types
```

Comment:
1. you are setting _options_ not a type
2. you do not need this cpt. Obvious here, the function name is pretty descriptive.